Add completeLoginViaTransaction and beginLoginViaTransaction functions in Frontend Client Package

jennifer-notifi opened the pull request:
Ported over `completeLoginViaTransaction` and `beginLoginViaTransaction` functions from `notifi-react-hooks` package into `notifi-frontend-client`

To Do: 
- Remove `walletPublicKey` in configuration and update `dappAddress` name to `tenantName`

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -1,12 +1,29 @@
 import { NotifiService } from '@notifi-network/notifi-graphql';
-import { UserFragmentFragment } from 'notifi-graphql/lib/gql/generated';
 
+import {
+  BeginLogInByTransactionResult,
+  CompleteLogInByTransactionInput,
+  CompleteLogInByTransactionMutation,
+  UserFragmentFragment,
+  WalletBlockchain,
+} from '../../notifi-graphql/lib/gql/generated';
```
we should not use relative paths, since this is a different NPM package

Can you explore why your VS Code isn't resolving via the NPM workspaces?

jennifer-notifi replied:
Ah, it does actually, just didn't realize that it was a best practice :)

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -1,12 +1,29 @@
 
 export type SignMessageFunction = (message: Uint8Array) => Promise<Uint8Array>;
 
+type BeginLoginProps = {
+  walletBlockchain: WalletBlockchain;
+  walletPublicKey: string;
+};
```
Omit<BeginLogInByTransactionInput, 'dappAddress'>`

jennifer-notifi replied:
Sweet, +1 to utility types: typescriptlang.org/docs/handbook/utility-types.html#omittype-keys

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -1,12 +1,29 @@
+  walletPublicKey: string;
+};
+
+type CompleteLoginProps = {
+  input: CompleteLogInByTransactionInput;
+};
```
`Omit<CompleteLogInByTransactionInput, 'dappAddress' | 'randomUuid'>`

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -1,12 +1,29 @@
+  input: CompleteLogInByTransactionInput;
+};
+
+let clientRandomUuid: string;
```
`let clientRandomUuid: string | null = null`

Also, this should be an instance variable, not a module scope variable

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +91,49 @@ export class NotifiFrontendClient {
+    const { transactionSignature } = input;
+    const { dappAddress } = this._configuration;
+
+    if (!clientRandomUuid) {
```
i prefer to totally avoid truthiness checks when possible. If we change this to `string | null` we can have an explicit `!== null`

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +91,49 @@ export class NotifiFrontendClient {
+    const { dappAddress } = this._configuration;
+
+    if (!clientRandomUuid) {
+      throw 'BeginLoginViaTransaction is required to be called first';
```
let's always throw an instance of error: `new Error('...')`

juni-notifi reviewed (changes requested):
Looks great! Just a few points of feedback:
- Please remove `.DS_Store` (add it to .gitignore?)
- In general, we should expose as much of the underlying graphql input params as we can
- For the SDK, we need much stricter type safety than we do in the Notifi Hub.

jennifer-notifi commented:
> Looks great! Just a few points of feedback:
> 
> * Please remove `.DS_Store` (add it to .gitignore?)
> * In general, we should expose as much of the underlying graphql input params as we can
> * For the SDK, we need much stricter type safety than we do in the Notifi Hub.

Thank you for the comments! I removed `DS_Store` using these steps: https://gist.github.com/lohenyumnam/2b127b9c3d1435dc12a33613c44e6308 and added it to the `.gitignore` file. I also reviewed what the DS_Store file: https://buildthis.com/ds_store-files-and-why-you-should-know-about-them/.

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -16,6 +29,7 @@ export class NotifiFrontendClient {
     private _configuration: NotifiFrontendConfiguration,
     private _service: NotifiService,
     private _storage: NotifiStorage,
+    private clientRandomUuid: string | null,
```
initialize the value? `= null`

jennifer-notifi replied:
Sorry I missed this, will update

jennifer-notifi replied:
Sorry for the miss. Will initialize.

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +88,49 @@ export class NotifiFrontendClient {
+
+    const nonce = result.beginLogInByTransaction.nonce;
+
+    this.clientRandomUuid = await window.crypto.randomUUID();
```
oooh ideally we can pull in a crypto dependency rather than depending on `window`

I know this is what happens in the current repo but it does feel wonky (this function will only be successful in a browser? i guess?)

jennifer-notifi replied:
After research on this, we concluded via Slack DM that using `crypto` alone does not rely on the browser being open: https://developer.mozilla.org/en-US/docs/Web/API/crypto_property

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +88,49 @@ export class NotifiFrontendClient {
+      dappAddress,
+    });
+
+    const nonce = result.beginLogInByTransaction.nonce;
```
these objects are nullable, but i don't see null-safe access

Can you share your typescript configuration with me? I think something's not right in the IDE

jennifer-notifi replied:
During our call, we confirmed the TS config is correct and the IDE is not advising on implementing chaining since the object cannot be nullable.

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +88,49 @@ export class NotifiFrontendClient {
+  }: CompleteLoginProps): Promise<CompleteLogInByTransactionMutation> {
+    const { dappAddress } = this._configuration;
+
+    if (this.clientRandomUuid == null) {
```
prefer explicit equivalence `===`

jennifer-notifi replied:
Ah, yes, explicit would be better

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +88,49 @@ export class NotifiFrontendClient {
+      walletAddress: walletAddress,
+      walletBlockchain: walletBlockchain,
+      dappAddress,
+      randomUuid: this.clientRandomUuid,
```
it's not an issue here, but in general, for async functions, it's good practice to store mutable variables in enclosing scope into a local `const` so the same value is used throughout the function's execution (between yields)

jennifer-notifi replied:
We mentioned during our call that it is best to call `const clientRandomUuid = this._clientRandomUuid;` and use `clientRandomUuid` throughout the function so no one can update it before the other parts of the function fire off.

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +81,61 @@ export class NotifiFrontendClient {
+
+    const nonce = result.beginLogInByTransaction.nonce;
+
+    if (nonce !== null) {
+      const ruuid = crypto.randomUUID();
+      this._clientRandomUuid = ruuid;
+      const encoder = new TextEncoder();
+      const data = encoder.encode(nonce + ruuid);
+
+      const hashBuffer = await crypto.subtle.digest('SHA-256', data);
+
+      const hashArray = Array.from(new Uint8Array(hashBuffer));
+      const logValue =
+        'Notifi Auth: 0x' +
+        hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
+      return { nonce: logValue };
+    } else throw new Error('Failed to begin login process');
```
nit: prefer early return pattern to avoid deep indentation 

```
if (nonce === null) {
  throw ...
}

...
```

jennifer-notifi replied:
good call - so the code doesn't have to find it at the end

jennifer-notifi replied:
> Nice, looks good to me.
> 
> Stupid question, but `npx lerna run build` completes with no errors, right? We should probably put that on a pre-commit hook somewhere.

<img width="1223" alt="Screen Shot 2022-08-18 at 11 16 39 AM" src="https://user-images.githubusercontent.com/108922885/185465812-e384041e-739c-4e25-a7a7-bb039d85421c.png">

Yes, it does! Let me look into adding the command in a pre-commit and how that would work.

juni-notifi commented on packages/notifi-frontend-client/lib/NotifiFrontendClient.ts:
```diff
@@ -73,4 +81,61 @@ export class NotifiFrontendClient {
+    transactionSignature,
+  }: CompleteLoginProps): Promise<Types.CompleteLogInByTransactionMutation> {
+    const { dappAddress } = this._configuration;
+    const clientRandomUuid = this._clientRandomUuid;
```
after this assignment, I think we should clear out `this._clientRandomUuid` -- what do you think?

jennifer-notifi replied:
Yes, I think that clearing it out makes sense in case `beginViaTransaction` has to run again in the same SDK instance.

```
const clientRandomUuid = this._clientRandomUuid;

this._clientRandomUuid = null;
```

juni-notifi reviewed (approved):
Nice, looks good to me.

Stupid question, but `npx lerna run build` completes with no errors, right? We should probably put that on a pre-commit hook somewhere.
